### DISCUSSION TITLE
Update tutorials

### DISCUSSION THREAD

**Discussion by KimberleyCook** [codebar/tutorials#485] [Announcements]
This is the start of a discussion with @brunogirin about updating the codebar tutorials

**Comment by brunogirin** [comment]
Here's the summary @KimberleyCook and I discussed so far.

# Aims

First, I would like to ensure we all agree on the aims of the tutorial update.

- Simplify maintenance by:
    - focusing on technologies we know well for which the tutorials will not require frequent updates,
    - pointing to external tutorials for the rest and to support our tutorials with additional content;
- Reduce barriers to entry by making it possible to get started without having to install anything on the student's laptop;
- Provide ways to go further through external tutorials;
- Bind everything in a coherent development story.

# Proposed lesson plan

This lesson plan attempts to keep as much of the current HTML+CSS tutorial as possible and then make the back-end tutorial work on top of it by adding more capabilities to the project. I moved some concepts to bonus lessons to streamline the core lessons while giving students the ability to dig deeper onto specific subjects. I also pointed at places where we could use an external tutorial in place of or to complement our own. This is not an exhaustive list, there may be more.

- Part 1: Front-end
    - Section 1: HTML and CSS
        - Lesson 1: Introducing HTML - simple page with HTML 5, headers, paragraphs, lists, etc
        - Lesson 2: Introducing CSS - simple styling
        - Lesson 3: Beyond the Basics - introduce links, images and the box model + more advanced CSS selectors
        - Lesson 4: Layouts - flexbox
        - Lesson 5: Responsive design basics
        - Lesson 6: Accessibility basics
        - Bonus lesson 1: Advanced HTML 5 (section, article and friends)
        - Bonus lesson 2: Advanced CSS 3 (complex selectors and media queries)
        - Bonus lesson 3: Colour (RBG, HSL, opacity, hex)
        - Bonus lesson 4: Introduction to SVG
        - External tutorials: CSS frameworks
    - Section 2: Javascript
        - Lesson 1: Dev tools + JS console
        - Lesson 2: number and string operations
        - Lesson 3: Functions
        - Lesson 4: Loops and arrays
        - Lesson 5: Classes
        - Lesson 6: Simple web component - [counter or something similar](https://meyerweb.com/eric/thoughts/2023/11/01/blinded-by-the-light-dom/)
        - Lesson 7: fetch
        - Lesson 8: Testing
        - External tutorials: JS frameworks
    - Section 3: running it all locally
        - Lesson 1: What is a web server?
        - Lesson 2: File management and relationship to URLs
- Part 2: Back-end
    - Section 1: Principles
        - Lesson 1: Why do you need a back-end?
    - Section 2: Python (+parallel tutorial on Ruby)
        - Lesson 1: Installation + intro to the REPL
        - Lesson 2: Number and string operations
        - Lesson 3: Functions
        - Lesson 4: Loops and lists
        - Lesson 5: Classes
        - Lesson 6: [Poetry](https://python-poetry.org/docs/) and project management
        - Lesson 7: Introduction to Flask (with support from [an external tutorial](https://blog.miguelgrinberg.com/post/the-flask-mega-tutorial-part-i-hello-world))
        - Lesson 8: Jinja2 templates - moving the Part 1 project to Flask (Sinatra for the Ruby version)
        - Bonus lesson 1: Introduction to REST APIs
        - External tutorials: other back-end Python frameworks ([Django](https://docs.djangoproject.com/en/5.0/intro/tutorial01/), FastAPI)
        - External tutorials: other back-end languages (PHP, Ruby, Java, Elixir, Node.JS, etc)
    - Section 3: Working with data
        - Lesson 1: HTML forms
        - Lesson 2: Processing form data in Flask
        - Lesson 3: More input types
    - Section 4: Databases
        - Lesson 1: What is a database?
        - Lesson 2: Introduction to SQL
        - Lesson 3: Creating tables and entering data
        - Lesson 4: Reading from tables
        - Lesson 5: SQLAlchemy - link the Flask back-end to the database (+external tutorial for details)
- Part 3: Professional development
    - Section 1: Command line
        - Lesson 1: Introduction to the command line
    - Section 2: Collaborative development
        - Lesson 1: Introduction to version control
        - Lesson 2: The git command line
        - Lesson 3: .gitignore and other files
        - Lesson 4: Uploading your project to Github
    - Section 3: Hosting
        - Lesson 1: Where does your code run?
        - Lesson 2: DNS basics
    - Section 4: Deployment
        - Lesson 1: Introduction to CI/CD
- Part 4: Other routes (external tutorials with guidance)
    - Section 1: Introduction to desktop apps
    - Section 2: Introduction to mobile apps
        - Lesson 1: Android
        - Lesson 2: iOS
        - Lesson 3: PWA
    - Section 3: Introduction to game development
    - Section 4: Introduction to data science

# Questions

Some questions I had when writing this document.

- Back-end:
    - Focus on Python only?
        - Suggestion: keep a Ruby version that follows the same path as Python
    - How do we deal with Python for back-end vs Python for data science?
    - Should we include Node.js?
        - Suggestion: no but we can link to external tutorials
    - Ensure the tutorials for the several languages follow the same structure
    - Simple operations
        - Strings
        - Built-in data structure (lists, dicts)
        - Code organisation in files
        - Functions
        - Classes
    - Point to external tutorials early
- The JavaScript case:
    - Introduce JS in the front-end using Web Components as it’s a W3C standard and is therefore stable? See [HTML web components](https://adactio.com/journal/20618)
    - Point to additional tutorials for vanilla JS?
    - Point to external tutorials for other frameworks e.g. React and Vue?
    - Point to external tutorials for Node.js and Express?
- Other subjects
    - Introduce the command line?
    - Introduce version control and git?

**Comment by mikejeuga** [comment]
This looks great @brunogirin . 
## Some suggestions:

### Part 2: Back-end:
 - [ ] We might need a consolidation step after the language construct. (so after lesson 5, a mini project would actually be helpful)
 - [ ] There should be a Testing step on the back-end
 - [ ] I think focusing on 1 language is enough (just personal opinion, also Python because [this](https://survey.stackoverflow.co/2024/technology#most-popular-technologies-language-learn).
 - [ ] I believe templating engines are a bit confusing for beginners, specially when they have React/JS in mind (I'd remove the jinja2 section). 
 - [ ] I feel like Part3, section 1 & 2 should actually come between FE & BE parts if the tutorial is sequential
 
### Questions:
  How do we deal with Python for back-end vs Python for data science?
  - [ ] Regardless, the student should first focus on Python only so Part2/Section2/Lessons 1 to 5
  - [ ] We should stress clearly that back-end development and data science are domains where Python is used. That way they can dissociate the tools from the domains. 
  - [ ] Regardless of the path, the student will have to learn:
    - [ ] Core Python
    - [ ] VCS
    - [ ] IDEs
  Only then they can be introduced to libraries that are specific to the domains (BE: Flask, FastAPI, etc || DS: Numpy, Pandas, etc).


My Opinion on the below:

Focusing on technologies we know well for which the tutorials will not require frequent updates:
  - [ ] It depends on what we are trying to achieve here. Are we looking to give student the tool to think outside of a specific tech or to become knowledgable about a specific set of technologies

pointing to external tutorials for the rest and to support our tutorials with additional content;
  - [ ] Agreed

Reduce barriers to entry by making it possible to get started without having to install anything on the student's laptop;
 - [ ] I think that might not work for everyone, I think it works well for some and for other that assume too little

Provide ways to go further through external tutorials;
  - [ ] Or give challenges so that students don't start tutorial after tutorial

Bind everything in a coherent development story.
  - [ ] This becomes hard if we do web and data science

**Reply by brunogirin** [reply]
> We might need a consolidation step after the language construct. (so after lesson 5, a mini project would actually be helpful)

Yes, that was the idea of lesson 8 in part 2, section 2 but we could have a mini-project after lesson 5 indeed.

> There should be a Testing step on the back-end

Yes, good point! Also Poetry makes this easy to set up.

> I think focusing on 1 language is enough (just personal opinion, also Python because [this](https://survey.stackoverflow.co/2024/technology#most-popular-technologies-language-learn).

I started with just Python but @KimberleyCook pointed out that the codebar web site is built in Ruby and is potentially a good project for people to contribute to. So it comes back to what we want the tutorials to be. One option would be to focus on Python but have a chapter on Ruby in the "other languages" lesson.

> Focusing on technologies we know well for which the tutorials will not require frequent updates:
> > It depends on what we are trying to achieve here. Are we looking to give student the tool to think outside of a specific tech or to become knowledgable about a specific set of technologies

Well, that's the first thing we need to decide on! If we want to help move the needle on diversity in tech, we need to give an onramp to finding a job in tech, for those who want. So when I created this, I had in mind that parts 1, 2 and 3 combined cover what someone would need to apply for a junior dev position at a company like mine.

> Reduce barriers to entry by making it possible to get started without having to install anything on the student's laptop;
> > I think that might not work for everyone, I think it works well for some and for other that assume too little

No indeed. My intention here was to make it possible, not to force it.

> Bind everything in a coherent development story.
> > This becomes hard if we do web and data science

Yes indeed. That's why I ended up moving data science to part 4.

**Comment by kodikos** [comment]
I think we need to know more about what's useful to our students. But something else I wanted to add is a page for resources, like virtual dev environments, and particularly code kata sites, like codewars, so they can practice what they learn.

**Reply by brunogirin** [reply]
Yes, totally agree with a page of resources! Regarding virtual dev environments, I toyed with the idea of adding a lesson on this but every time I think about the subject, I over-complicate it so I am open to any suggestions on the subject.

**Comment by brunogirin** [comment]
# Discussion at uncodebar 10 on 21 Sep 2024

## Needs

### Student

What do you expect from codebar?
- Move to a career in tech - web based development
- Move to a career in data analytics
- Curious, don’t know what coding is
- Build on skills you already have
- Tech meetup, social, community
- Break barrier from learning on your own, ask people
- Transfer from one programming language to another one
- How to analyse problems
- Cloud technologies

### Maintainers

- Easy to maintain

## Suggestions
- Make NodeJS a first class option?
- Java?
- Ditch Ruby!
- Personal website track? + personal portfolio
- Understand your configuration + install debug
- Explain OO vs functional, how to switch to another language, etc
- How to take people out of HTML+CSS?
- Guidelines to do a project
- How to choose the right tool for the job
- Screenshots on the way + what it looks like at the end
- Developer tools
- Setup code editor + tools
- Debugging tools
- Testing

## Governance

- #tutorials channel on Slack
- Pull requests
- Clear definition of content
- Who is in charge of the tutorials?
- Status updates
- Subjects of the tutorials (owls and stuff)?

**Comment by brunogirin** [comment]
# Discussion during Organiser meeting on Thursday 26th September

At the end of relevant lessons, add a "Your Turn" section where we suggest ways the student could improve or amend the project:
- Don't give any solution, just pointers about what they can do
- Word it in a similar way to possible job interview assignments
- Ensure it focuses on what they've just learnt

Suggestions not discussed during the meeting:
- For end of lesson, include a small and simple exercise that focuses on the lesson in question
- For the end of section or part, make it more complex such that it uses everything that was covered, potentially mini-projects for the largest

@mikejeuga @KimberleyCook please add if you think I missed anything important that we discussed

**Comment by brunogirin** [comment]
# Part 0: orientation

Would an initial orientation part be useful? Possible scope of such a part:
- Orientation around the tutorial, what parts you might want to focus on based on your aims
- "Know Your Computer" section that discusses the conceps of files, folders, etc
- Tooling section to introduce development tools and anything you might need to follow the tutorials

**Comment by brunogirin** [comment]
# Status update

As discussed with @KimberleyCook , the first step is to upgrade to a version of Ruby, Jekyll and JS libraries that are reasonably recent. I started looking into it then got distracted by my house move so will get back to it as soon as I am able.
